freqtrade/exchange/exchange.py

```diff
@@ -2018,102 +2178,232 @@ def _build_ohlcv_dl_jobs(
+            """
+            Refresh in-memory OHLCV asynchronously and set `_trades` with the result
+            Loops asynchronously over pair_list and downloads all pairs async (semi-parallel).
+            Only used in the dataprovider.refresh() method.
+            :param pair_list: List of 2 element tuples containing pair, interval to refresh
+            :param since_ms: time since when to download, in milliseconds
+            :param cache: Assign result to _trades. Usefull for one-off downloads like for pairlists
+            :param drop_incomplete: Control candle dropping.
+                Specifying None defaults to _ohlcv_partial_candle
+            :return: Dict of [{(pair, timeframe): Dataframe}]
+            """
+            logger.debug("Refreshing candle (OHLCV) data for %d pairs", len(pair_list))
+
+            # Gather coroutines to run
+            input_coroutines, cached_pairs = self._build_ohlcv_dl_jobs(pair_list, since_ms, cache)
+
+            results_df = {}
+            # Chunk requests into batches of 100 to avoid overwelming ccxt Throttling
+            for input_coro in chunks(input_coroutines, 100):
+
+                async def gather_stuff():
+                    return await asyncio.gather(*input_coro, return_exceptions=True)
+
+                with self._loop_lock:
+                    results = self.loop.run_until_complete(gather_stuff())
+
+                for res in results:
+                    if isinstance(res, Exception):
+                        logger.warning(f"Async code raised an exception: {repr(res)}")
+                        continue
+                    # Deconstruct tuple (has 5 elements)
+                    pair, timeframe, c_type, ticks, drop_hint = res
+                    drop_incomplete = drop_hint if drop_incomplete is None else drop_incomplete
+                    # TODO: here ohlcv candles get saved into self._trades
+                    ohlcv_df = self._process_ohlcv_df(
+                        pair, timeframe, c_type, ticks, cache, drop_incomplete)
+
+                    results_df[(pair, timeframe, c_type)] = ohlcv_df
+
+            # Return cached trades
+            for pair, timeframe, c_type in cached_pairs:
+                results_df[(pair, timeframe, c_type)] = self.klines(
+                    (pair, timeframe, c_type),
+                    copy=False
+                )
+
+            return results_df
+
+    def needed_candle_ms(self, timeframe:str, candle_type:CandleType):
+        one_call = timeframe_to_msecs(timeframe) * self.ohlcv_candle_limit(
+            timeframe, candle_type)
+        move_to = one_call * self.required_candle_call_count
+        now = timeframe_to_next_date(timeframe)
+        return int((now - timedelta(seconds=move_to // 1000)).timestamp() * 1000)
+
+    def refresh_latest_trades(self, 
+                              pair_list: ListPairsWithTimeframes ,
+                              data_handler: Callable,# IDataHandler, 
```

Comment:
Callable is for sure wrong - as you're not passing a callable in here, but a class instance (i've looked at where you actually call this).
not sure why it can't be IDataHandler though?

Reply:
I will try again with `IDataHandler`

Reply:
I get a circular import error trying to use `IDataHandler`

```
ImportError: cannot import name 'timeframe_to_seconds' from partially initialized module 'freqtrade.exchange' (most likely due to a circular import) (/home/joe/zk/efforts/ft/freqtrade/freqtrade/exchange/__init__.py)
```